warehouse/email/__init__.py

```diff
@@ -207,6 +207,11 @@ def send_password_compromised_email_hibp(request, user):
     return {}
 
 
+@_email("token-compromised-leak", allow_unverified=True)
+def send_token_compromised_email_leak(request, user, *, public_url, origin):
+    return {"username": user.username, "public_url": public_url, "origin": origin}
```

Comment:
Looks like this fails when sent from within a worker, because rendering a template uses `request.timings` that is set in `on_new_request` and apparently not set on the request object that is received by the task.

```
worker_1         | Traceback (most recent call last):
worker_1         |   File "/opt/warehouse/lib/python3.8/site-packages/celery/app/trace.py", line 412, in trace_task
...
worker_1         |   File "/opt/warehouse/src/warehouse/integrations/github/tasks.py", line 20, in analyze_disclosure_task
worker_1         |     utils.analyze_disclosure(
worker_1         |   File "/opt/warehouse/src/warehouse/integrations/github/utils.py", line 333, in analyze_disclosure
worker_1         |     _analyze_disclosure(
worker_1         |   File "/opt/warehouse/src/warehouse/integrations/github/utils.py", line 322, in _analyze_disclosure
worker_1         |     send_token_compromised_email_leak(
worker_1         |   File "/opt/warehouse/src/warehouse/email/__init__.py", line 143, in wrapper
worker_1         |     msg = EmailMessage.from_template(name, context, request=request)
worker_1         |   File "/opt/warehouse/src/warehouse/email/services.py", line 43, in from_template
worker_1         |     subject = render(f"email/{email_name}/subject.txt", context, request=request)
worker_1         |   File "/opt/warehouse/lib/python3.8/site-packages/pyramid/renderers.py", line 75, in render
worker_1         |     result = helper.render(value, None, request=request)
worker_1         |   File "/opt/warehouse/lib/python3.8/site-packages/pyramid/renderers.py", line 469, in render
worker_1         |     registry.notify(system_values)
worker_1         |   File "/opt/warehouse/lib/python3.8/site-packages/pyramid/registry.py", line 109, in notify
worker_1         |     [_ for _ in self.subscribers(events, None)]
worker_1         |   File "/opt/warehouse/lib/python3.8/site-packages/zope/interface/registry.py", line 448, in subscribers
worker_1         |     return self.adapters.subscribers(objects, provided)
worker_1         |   File "/opt/warehouse/lib/python3.8/site-packages/zope/interface/adapter.py", line 619, in subscribers
worker_1         |     subscription(*objects)
worker_1         |   File "/opt/warehouse/lib/python3.8/site-packages/pyramid/config/adapters.py", line 101, in derived_subscriber
worker_1         |     return subscriber(arg[0])
worker_1         |   File "/opt/warehouse/src/warehouse/metrics/event_handlers.py", line 77, in on_before_render
worker_1         |     timings = request.timings
worker_1         | AttributeError: 'Request' object has no attribute 'timings'
```

@di @ewdurbin any idea how I should be solving that ?

Reply:
Hmm, I guess we haven't sent any emails from worker tasks before.

I think we can do the same thing we're doing here:
https://github.com/pypa/warehouse/blob/ea21a674b6165d61340ae976da67a08729902f6a/warehouse/metrics/event_handlers.py#L48

Here:

https://github.com/pypa/warehouse/blob/ea21a674b6165d61340ae976da67a08729902f6a/warehouse/tasks.py#L88-L95

Although I'm kind of expecting you to run into other things missing on the request as well. But maybe not!

Reply:
Seems to make sense, I'll try it ! Thank you !

Reply:
Trying to see if it makes more sense architecturally to copy-paste code (3 lines) or to import code from metrics in tasks (which seems like a bad idea). I'm going to go with the copy-paste, but feel to tell me if you think it's not the best.

This tends to make me think the timings part should be in independent from the metrics part. The timing should ideally be attached to the common code that prepares a request object, whether that be a HTTP request or a Celery request. But that's not something that makes sense to tackle in this PR.

Reply:
> Although I'm kind of expecting you to run into other things missing on the request as well. But maybe not!

Good call !

`Task warehouse.email.send_email[c411274d-7b88-4944-9256-af118116dce7] retry: Retry in 180s: IntegrityError('(psycopg2.errors.NotNullViolation) null value in column "ip_address" violates not-null constraint`

Reply:
Looks like we need to add something for `request.remote_addr` (maybe `127.0.0.1`) or make this field nullable. Probably makes more sense for it to be null since there isn't actually an IP generating this.

Reply:
(That can be a separate PR btw)

Reply:
Haha, was about to say the same :)

Reply:
Ok, we're good now. Used `"127.0.0.1"` like you said, but "N/A" could have been good too. Next step will be to make this nullable \o/